### PR TITLE
Base CSV Writer for MCO Events

### PR DESCRIPTION
**Pull request by Corwinpro**
This PR approaches [this issue](https://github.com/force-h2020/force-bdss-plugin-itwm-example/issues/60) and attempts to generalise the `CSVWriter` functionality

### Summary

`CSVWriter` is a notification listener that logs the MCO data during runtime, in `.csv` format.

### Done
- Refactored `BaseDriverEvent.__getstate__`: now, any `trait` and `List(trait)` objects will be automatically parsed by the base class using `nested_getstate` method. **This breaks backwards compatibility**: child classes no longer need to implement `__getstate__` for simple traits and lists of traits.
- Created methods to generate `MCOStartEvent` and `MCOFinishEvent` instances on the `BaseMCOModel` class. These methods are then referred to in `OptimizeOperation._deliver_start_event` and `OptimizeOperation._deliver_finish_event`, allowing MCO-specific driver events to be contributed by the `BaseMCOModel`.
- Implemented `BaseCSVWriter`. The `deliver` method parses incoming `event: Union(MCOStartEvent, MCOProgressEvent)` object and creates header / row data. This data is then written to file. If a specific header / row format is required, we suggest that the user implements their own `parse_start_event` and `parse_progress_event`, or extends them with additional logic.
- Developed dict-style design for `row_data`. Now, when a header is created with `MCOStartEvent`, a new `row_data` dictionary is instantiated, with no data. Data coming from `MCOProgressEvent` is assigned to `row_data` entries such that the first data value from the event corresponds to the first header value, and so on. This means that the `MCOProgressEvent` will fill out the first available `header` slots.

### Possible ToDos

- (see #247) Now the `MCOModel` is responsible for creating (custom) `MCOStart` and `MCOFinish` events. It seems reasonable to
-- Implement `MCOModel.create_progress_event` method which handles the optimal points / optimal KPIs broadcasting, and
-- Refactor the `BaseMCO.notify_new_point` and `BaseMCO.notify` to `BaseMCOModel`. Then it would be much easier to broadcast the progress events since the model is responsible for their initialisation.

### CODE REVIEW

File: force_bdss/core_driver_events.py [hunk 1]
```diff
@@ -6,27 +6,51 @@
...
+        for key in state:
+            # We safely attempt to get state of the zero and first
+            # level objects of the object state dictionary
+            try:
+                if isinstance(state[key], (tuple, list)):
+                    state[key] = [el.__getstate__() for el in state[key]]
+                else:
+                    state[key] = state[key].__getstate__()
+            except AttributeError:
+                pass
```
[the remark this answers is not in the document]
Comment (Corwinpro) [reply force_bdss/core_driver_events.py]: Thank you, will do. So far I can only see how to `__getstate__` one level in depth.

File: force_bdss/core_driver_events.py [hunk 2]
```diff
@@ -37,8 +61,19 @@ class MCOProgressEvent(BaseDriverEvent):
...
+        `force_wfmanager.wfmanager_setup_task._server_event_mainthread`
+        Can we refactor this?
+
+        Warning: `weights` attribute is NOT serialized here. We expect it to be
+        refactored to custom MCOProgressEvent class.
+
+        Returns:
+            List(Datavalue.value): values of the event optimal points and kpis
+        """
+        event_datavalues = self.optimal_point + self.optimal_kpis
```
[the remark this answers is not in the document]
Comment (Corwinpro) [reply force_bdss/core_driver_events.py]: - My first impression was that for custom `MCOProgressEvent`s one should a) implement additional logic for `serialize` if needed, b) call `super().serialize()` and therefore access the base implementation of  the `MCOProgressEvent.serialize`.

- I should've raised an issue in the wfmanager repo for that. I will create one if I hadn't.

[the remark this answers is not in the document]
Comment (flongford) [reply force_bdss/core_driver_events.py]: > * My first impression was that a custom `MCOProgressEvent` should a) implement additional logic for `serialize` if needed, b) call `super().serialize()` and therefore access the base implementation of  the `MCOProgressEvent.serialize`.  See [itwm example](https://github.com/force-h2020/force-bdss-plugin-itwm-example/blob/fd3ca69671dc5c546247f6798cdbf777834468d6/itwm_example/csv_writer/csv_writer.py#L7)

Yes, that works too - perhaps we can include a very simple example in the `serialize` docsting to demonstrate this?

[the remark this answers is not in the document]
Comment (Corwinpro) [reply force_bdss/core_driver_events.py]: In fact now it seems that `CSVWriter` should not be responsible for this kind of data addition / manipulation. I would say even that both the `parse_progress_event` and `parse_start_event` should only access the serialised event data via `event.serialize()`.

[the remark this answers is not in the document]
Comment (Corwinpro) [reply force_bdss/core_driver_events.py]: This also presumes that we should be able to supply custom MCOStart and MCOProgress events during the MCO run.

[the remark this answers is not in the document]
Comment (flongford) [reply force_bdss/core_driver_events.py]: Yes, that makes sense, feels much cleaner

[the remark this answers is not in the document]
Comment (Corwinpro) [reply force_bdss/core_driver_events.py]: I will think about this. Seems to be fairly straightforward for a progress event, while the start event is hardcoded into bdss.

[the remark this answers is not in the document]
Comment (Corwinpro) [reply force_bdss/core_driver_events.py]: We refactored the `MCOStartEvent` and `MCOFinishEvent` to `BaseMCOModel`.

**Comment by flongford** [comment]
> -- Refactor the `BaseMCO.notify_new_point` and `BaseMCO.notify` to `BaseMCOModel`. Then it would be much easier to broadcast the progress events since the model is responsible for their initialisation.

It may be more appropriate for the `Workflow` class to handle this, since then the `OptimizeOperation` can just listen to a single location for any events being broadcast by both `BaseMCO*` or `BaseDataSource*` objects.

**Pull request merged**